src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,329 @@
+     * length inclusive of structural overhead.
+     *
+     * @see https://docs.mongodb.com/manual/reference/limits/#Index-Key-Limit
+     *
+     * CAUTION: This store relies on all client and server nodes to have
+     * synchronized clocks for lock expiry to occur at the correct time.
+     * To ensure locks don't expire prematurely; the lock TTL should be set
+     * with enough extra time to account for any clock drift between nodes.
+     * @see self::createTTLIndex()
+     *
+     * writeConcern, readConcern and readPreference are not specified by
+     * MongoDbStore meaning the collection's settings will take effect.
+     * @see https://docs.mongodb.com/manual/applications/replication/
+     */
+    public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)
+    {
+        if (!isset($options['database'])) {
+            throw new InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+        ), $options);
+
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * Create a TTL index to automatically remove expired locks.
+     *
+     * This should be called once during database setup.
+     *
+     * Alternatively the TTL index can be created manually:
+     *
+     *  db.lock.ensureIndex(
+     *      { "expires_at": 1 },
+     *      { "expireAfterSeconds": 0 }
+     *  )
+     *
+     * Please note, expires_at is based on the application server. If the
+     * database time differs; a lock could be cleaned up before it has expired.
+     * Set a positive expireAfterSeconds to account for any time drift between
+     * application and database server.
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     *
+     * @return string The name of the created index
+     *
+     * @throws UnsupportedException          if options are not supported by the selected server
+     * @throws MongoInvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException        for other driver errors (e.g. connection errors)
+     */
+    public function createTTLIndex(): string
+    {
+        $keys = array(
+            'expires_at' => 1,
+        );
+
+        $options = array(
+            'expireAfterSeconds' => 0,
+        );
+
+        return $this->getCollection()->createIndex($keys, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockStorageException if failed to save to storage (fallback exception)
+     * @throws LockExpiredException when save is called on an expired lock
+     */
+    public function save(Key $key)
+    {
+        $token = $this->getUniqueToken($key);
+        $now = microtime(true);
+
+        $filter = array(
```

Comment:
No;

We might want to pickup a leftover document from a previous instantiation of `Key` (same resource / `_id`) with a different `token`.

OR

We can overwrite our instantiation of `Key`'s lock (`AbstractStoreTest::testSaveTwice`).

Reply:
I also don't understand the design:

```
We might want to pickup a leftover document from a previous instantiation of Key (same resource / _id) with a different token
```
How will the new process know that the lock resource is free to be acquired in a document with the key id already exists?

Reply:
> IMO every store should clean up the key as well as the associated data with it on a release or expiration

We do this on the MongoDBStore via the usage of a MongoDB TTL Index (see the phpdoc on the `__construct` and also https://github.com/symfony/symfony-docs/pull/9807/files#diff-1dc6462c0bc00fcdc7cba99d2f13e400R484

The implementation however is designed to be resilient in the event someone doesn't create a MongoDb TTL Index. It will begin to recycle documents ONLY if they are expired or they are owned by the current thread (this is why we have an `$or` in the query).

e.g.

```php
$k1 = new \Symfony\Component\Lock\Key('my-resource');
$k2 = new \Symfony\Component\Lock\Key('my-resource'); // This has a different "token" to k1

$store->save($k1);

$store->save($k1);
// This is ok, we are re-saving k1 (scenario enforced by `AbstractStoreTest::testSaveTwice`)

$store->save($k2);
// This is NOT ok, k1 has the lock

sleep($lockTtl + 1);
// $k1 has now expired

$store->save($k2);
// This is ok, re-cycling the document created by k1 IF the use ignored the recommendation to create a MongoDb TTL Index
```

Reply:
Okay this is making more sense to me, thanks for the explanation. Let me know if I am understanding this wrong:

In your above case, it is assumed that the client created a TTL index on the collection. There is a possibility that user doesn't explicitly call this. So will you run into the following scenario?
```
// TTL is not set on the collection
- $store->save($k1);  // it somehow creates the document with an expiration time in future, by default it is 300 seconds 
- // the client forgets to call release and the thread finishes processing
- // the collection still has `k1` with the original $token because the expiration value has no effect as TTL index was never created on Mongo DB through Database administration or by the client during DB setup once initially
- $store->save($k2); // this should never work because the token associated with `_id` never expires because no TTL Index was created
```

Or does not setting a TTL Index on the collection mean that, the document itself won't be deleted but the data associated with `_id` (`$token` in our case) will get deleted but the document itself remains?

Reply:
My scenario above does NOT assume a TTL index was created. The PHP will behave the same regardless of if the TTL Index exists. The only difference is if a document is left behind because PHP crashed and didn't get to remove the lock, the lock / document would persist forever (or until the next time another lock was acquired with the same resource id).

Think of a TTL Index like a cron that runs on an SQL database periodically:

```sql
DELETE FROM lock WHERE expires_at < (NOW() + :expireAfterSeconds);
```

It simply cleans up expired locks

Assuming a TTL index is NOT created...

> the collection still has `k1` with the original $token because the expiration value has no effect as TTL index was never created on Mongo DB through Database administration or by the client during DB setup once initially

This is true, a `document` (a.k.a `row` in RDBMS world) will remain in the `lock` collection (a.k.a. `table` in RDBMS world) indefinitely.

> $store->save($k2); // this should never work because the token associated with `_id` never expires because no TTL Index was created

This will work as soon as the `expires_at <= $now` ($k2 will overwritten into `$k1`'s expired document), `MongoDbStore::save` will:
```php
$filter = array(
    '_id' => (string) $key,
    '$or' => array(
        array(
            'token' => $token,
        ),
        array(
            'expires_at' => array(
                '$lte' => $this->createDateTime($now),
            ),
        ),
    ),
);

$update = array(
    '$set' => array(
        '_id' => (string) $key,
        'token' => $token,
        'expires_at' => $this->createDateTime($now + $this->initialTtl),
    ),
);

$options = array(
    'upsert' => true,
);
```

If however you tried to `$store->save($k2);` BEFORE `expires_at <= $now` (e.g. before `$k1` has expired), the upsert would effectively attempt to `INSERT` which would cause a `Mongo error E11000 - DuplicateKey`.

Reply:
Okay makes sense now.